src/C++/SSLSocketInitiator.h

```diff
@@ -170,11 +175,15 @@ class SSLSocketInitiator : public Initiator, SocketConnector::Strategy
   void onDisconnect( SocketConnector&, socket_handle);
   void onError( SocketConnector& );
   void onTimeout( SocketConnector& );
-  bool handshakeSSL(SSL* ssl);
+  void disconnectPendingSSLHandshakesThatTakeTooLong(time_t now);
+  SSLHandshakeStatus handshakeSSL(SSLSocketConnection* connection);
+  void handshakeSSLAndHandleConnection(SocketConnector& connector, socket_handle s);
   void getHost( const SessionID&, const Dictionary&, std::string&, short&, std::string&, short& );
 
   SessionToHostNum m_sessionToHostNum;
   SocketConnector m_connector;
+
+  SocketConnections m_pendingSSLHandshakes;
```

Comment:
Added a new state where pendingConnections will be at until sslHandshake() succeeds or fails.